### ISSUE TITLE
Is Schema<->Context compare working correctly?

### ISSUE THREAD

**Issue by CaiusJard** [ErikEJ/EFCorePowerTools#1118]
Just thought I'd ask a few Qs because I'm not sure how this is supposed to appear..

So, I ran a compare of by context vs my test db, because I want to make sure that my test DB is up to date relative to live; my context was scaffolded from live using scaffold-dbcontext..

The first time I ran it I got a few diffs.. I eventually worked out what it meant - none of the child ends of my FKs had indexes hence the large number of complaints about indexes "Expected blahblah, Actual (blank)"

![image](https://user-images.githubusercontent.com/68939130/134538812-fdaf1696-f007-4cc9-b4e3-6515fee36c9f.png)

There's also a typo in an index name, probably not a train smash but I'll fix it anyway..

The NewStatus and OldStatus were interesting - in code theyre enums, but I store my enums as strings, so the enum props have an attribute that makes EF store them as string. In code they aren't nullable, because theyre enums but in the DB they are. I should probably sort that out...

I didn't actually fix this/make any changes in relation to it, but I've noted that when I re-run the compare these things don't seem to be "differences":

![image](https://user-images.githubusercontent.com/68939130/134539393-f7aa81b9-bb06-4aea-9066-11adca099c3a.png)

Unticking "show diffs" causes them to appear:

![image](https://user-images.githubusercontent.com/68939130/134539673-bbb0a369-56b4-4879-a500-073a1a5dbd0c.png)

---

So in summary:

* Is this null/notnull is not deemed to be a difference for "show diffs" purposes a bug?
* I've never seen anything in any of the Type/State/Attribute columns - is that expected?
* Would it be more logical to have "expected" and "found" be shown as the same value if they aren't different?

### Further technical details
EF Core Power Tools version: 2.5
Database engine: SQLS2017
Visual Studio version: 16.11.2

**Comment by ErikEJ** [comment]
Good questions, let's ask @jonpsmith, who created the underlying diff tool

**Comment by JonPSmith** [comment]
Hi @CaiusJard,

As @ErikEJ  states the EFCorePowerTools tool uses my [EfCore.SchemaCompare](https://github.com/JonPSmith/EfCore.SchemaCompare) library. This library returns errors only, in a string format - for instance

```
"DIFFERENT: MyEntity->Property 'MyString'. Expected = varchar(max), found = nvarchar(max)"
```

EFCorePowerTools felt that a full list of the compares could be useful so it uses the logs that EfCore.SchemaCompare uses to handle a two-part compare (1st part compares matching property/columns, 2nd part looks for extra / missing tables, columns etc.)

To answer your questions:

> Is this null/notnull is not deemed to be a difference for "show diffs" purposes a bug?

EfCore.SchemaCompare will return an error if the EF Core Model the database has a different null setting to a database. Its a difference and can cause problems (e.g. database column is NOT NULL, and EF Core property is null and you send a null value).

> I've never seen anything in any of the Type/State/Attribute columns - is that expected?

I'm not sure what you means here. Are you referring the Type/State/Attribute data in the logs? If so I can say that it only sets certain parts of the log if there is an error.

> Would it be more logical to have "expected" and "found" be shown as the same value if they aren't different?

It might be, but the focus is on compare errors.

**Comment by ErikEJ** [comment]
Thanks, @JonPSmith
